lib/charms/bind/v0/dns_record.py

```diff
@@ -0,0 +1,570 @@
+            relation_data: the relation data.
+
+        Returns:
+            A DNSRecordRequirerData instance.
+
+        Raises:
+            ValidationError if the value is not parseable.
+        """
+        try:
+            loaded_data = {}
+            for key, value in relation_data.items():
+                if value:
+                    loaded_data[key] = json.loads(value)
+            return cls(**loaded_data)
+        except json.JSONDecodeError as ex:
+            raise ValidationError.from_exception_data(ex.msg, [])
+
+
+class DNSRecordRequestProcessed(ops.RelationEvent):
+    """DNS event emitted when a new request is processed.
+
+    Attributes:
+        dns_record_provider_relation_data: the DNS provider relation data.
+        dns_domains: list of processed domains.
+        dns_entries: list of processed entries.
+    """
+
+    @property
+    def dns_record_provider_relation_data(self) -> DNSRecordProviderData:
+        """Get a DNSRecordProviderData for the relation data."""
+        assert self.relation.app, "DNS record relation data accessed before relation setup."
+        return DNSRecordProviderData.from_relation_data(self.relation.data[self.relation.app])
+
+    @property
+    def dns_domains(self) -> Optional[List[DNSProviderData]]:
+        """Fetch the DNS domains from the relation."""
+        return self.dns_record_provider_relation_data.dns_domains
+
+    @property
+    def dns_entries(self) -> Optional[List[DNSProviderData]]:
+        """Fetch the DNS entries from the relation."""
+        return self.dns_record_provider_relation_data.dns_entries
+
+
+class DNSRecordRequestReceived(ops.RelationEvent):
+    """DNS event emitted when a new request is made.
+
+    Attributes:
+        dns_record_requirer_relation_data: the DNS requirer relation data.
+        dns_domains: list of requested domains.
+        dns_entries: list of requested entries.
+    """
+
+    @property
+    def dns_record_requirer_relation_data(self) -> DNSRecordRequirerData:
+        """Get a DNSRecordRequirerData for the relation data."""
+        assert self.relation.app
+        return DNSRecordRequirerData.from_relation_data(self.relation.data[self.relation.app])
+
+    @property
+    def dns_domains(self) -> Optional[List[RequirerDomain]]:
+        """Fetch the DNS domains from the relation."""
+        return self.dns_record_requirer_relation_data.dns_domains
+
+    @property
+    def dns_entries(self) -> Optional[List[RequirerEntry]]:
+        """Fetch the DNS entries from the relation."""
+        return self.dns_record_requirer_relation_data.dns_entries
+
+
+class DNSRecordRequiresEvents(ops.CharmEvents):
+    """DNS record requirer events.
+
+    This class defines the events that a DNS record requirer can emit.
+
+    Attributes:
+        dns_record_request_processed: the DNSRecordRequestProcessed.
+    """
+
+    dns_record_request_processed = ops.EventSource(DNSRecordRequestProcessed)
+
+
+class DNSRecordRequires(ops.Object):
```

Comment:
Shouldn't we provide as part of the public API here method like: `add_domain`, `remove_domain`, `update_domain` and the same for records ?

Reply:
Yes, Ibut this is the minimal functionality required for the relation